### PR TITLE
Add support for storing model in file like object.

### LINKED ISSUE #86: Model can only be stored to files.
What if I want to store a model in a database? See Pull Request #85 for a fix. Sorry for first opening the pull and then the issue!

### PR DESCRIPTION
**Pull request by erniejunior**
Sometimes it is useful to store a model somewhere else than in a file. 
My use case is storing the model in an 'attachment' of the [hyperopt](https://github.com/hyperopt/hyperopt) library or storing it in the [GridFS](https://docs.mongodb.com/manual/core/gridfs/) of a MongoDB. 

Fixes #86

### CODE REVIEW

File: stable_baselines/common/base_class.py [hunk 1]
```diff
@@ -177,7 +177,7 @@ def save(self, save_path):
         """
         Save the current parameters to file
 
-        :param save_path: (str) the save location
+        :param save_path: (str or file-like) the save location
```
Comment (araffin) [inline stable_baselines/common/base_class.py]:
```suggestion
        :param save_path: (str or file-like object) the save location
```

File: stable_baselines/common/base_class.py [hunk 2]
```diff
@@ -198,23 +198,29 @@ def load(cls, load_path, env=None, **kwargs):
...
-        if ext == "":
-            save_path += ".pkl"
+        if isinstance(save_path, str):
+            _, ext = os.path.splitext(save_path)
+            if ext == "":
+                save_path += ".pkl"
 
-        with open(save_path, "wb") as file:
-            cloudpickle.dump((data, params), file)
+            with open(save_path, "wb") as file:
```
Comment (araffin) [inline stable_baselines/common/base_class.py]:
```suggestion
            with open(save_path, "wb") as file_:
```

`file` is a python keyword. (I know that was there before but I think that is the right moment to change that ;) )

File: stable_baselines/common/base_class.py [hunk 3]
```diff
@@ -198,23 +198,29 @@ def load(cls, load_path, env=None, **kwargs):
...
+                cloudpickle.dump((data, params), file)
+        else:
+            cloudpickle.dump((data, params), save_path)
```
Comment (araffin) [inline stable_baselines/common/base_class.py]:
```suggestion
            # Here save_path is a file-like object, not a path
            cloudpickle.dump((data, params), save_path)
```

File: stable_baselines/common/base_class.py [hunk 4]
```diff
@@ -198,23 +198,29 @@ def load(cls, load_path, env=None, **kwargs):
...
+                    load_path += ".pkl"
+                else:
+                    raise ValueError("Error: the file {} could not be found".format(load_path))
 
-        with open(load_path, "rb") as file:
-            data, params = cloudpickle.load(file)
+            with open(load_path, "rb") as file:
+                data, params = cloudpickle.load(file)
+        else:
+            data, params = cloudpickle.load(load_path)
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: Missing comment (cf above)

File: tests/test_save.py [hunk 5]
```diff
@@ -50,12 +57,23 @@ def test_model_manipulation(model_class):
         acc_reward = sum(acc_reward) / N_TRIALS
 
         # saving
-        model.save("./test_model")
+        if storage_method == "path":
+            model.save("./test_model")
+        else:
+            b_io = BytesIO()
```
Comment (araffin) [inline tests/test_save.py]: Missing comment

**Comment by erniejunior** [comment]
Yes we could check if the input is file-like. But the check for the string is more reliable I guess? According to [this](https://stackoverflow.com/questions/1661262/check-if-object-is-file-like-in-python) we should not check if the object is file-like and rather catch the exceptions that get raised when the input is wrong.

**Comment by araffin** [comment]
I agree with you, we should avoid breaking API for now, maybe for the next major release.
So I would leave a note or TODO next to it, stating that this should be renamed in future version.

I read the stack overflow thread, and I also think that are too many potential objects. So, yes, let's leave it like that for now.

File: tests/test_save.py [hunk 6]
```diff
@@ -19,10 +20,16 @@
     TRPO,
 ]
 
+STORE_METHOD = [
```
Comment (araffin) [inline tests/test_save.py]:
```suggestion
STORE_METHODS = [
```

File: tests/test_save.py [hunk 7]
```diff
@@ -19,10 +20,16 @@
...
+    "path",
+    "file-like"
+]
+
 
 @pytest.mark.slow
 @pytest.mark.parametrize("model_class", MODEL_LIST)
-def test_model_manipulation(model_class):
+@pytest.mark.parametrize("storage_method", STORE_METHOD)
```
Comment (araffin) [inline tests/test_save.py]:
```suggestion
@pytest.mark.parametrize("storage_method", STORE_METHODS)
```

File: stable_baselines/common/base_class.py [hunk 8]
```diff
@@ -198,23 +198,31 @@ def load(cls, load_path, env=None, **kwargs):
...
+                    load_path += ".pkl"
+                else:
+                    raise ValueError("Error: the file {} could not be found".format(load_path))
 
-        with open(load_path, "rb") as file:
-            data, params = cloudpickle.load(file)
+            with open(load_path, "rb") as file:
+                data, params = cloudpickle.load(file)
+        else:
+            # Here save_path is a file-like object, not a path
```
Comment (araffin) [inline stable_baselines/common/base_class.py]:
```suggestion
            # Here load_path is a file-like object, not a path
```

**Comment by erniejunior** [comment]
Any action required from my side now? (I have not used the GitHub pull request feature in such detail before)

**Comment by araffin** [comment]
I was waiting for both CI build to finish and @hill-a approval.
I will merge it tomorrow if he is still busy after a final check ;)

**Comment by araffin** [comment]
merged, thank you for your contribution (I will add it to the changelog)

**Pull request merged**